File: backend/internal/models/canary.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// CanaryToken is an administrator-created tracking link ("canary" / honeytoken).
// When the public slug is visited the server records the visitor's real IP,
// User-Agent and other request metadata, then redirects (HTTP 302) to a benign
// TargetURL chosen by the administrator.
//
// Defensive use cases: detecting when a planted bait document / link is opened,
// deanonymising who is probing leaked credentials, and phishing-back
// investigations on authorised engagements.
//
// Ethics / authorisation: this records PII (IP address, User-Agent). It must
// only be operated inside an authorised investigation, and the redirect target
// must be a harmless page. The UI surfaces this warning to the operator.
type CanaryToken struct {
	ID uuid.UUID `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`

	// Slug is the short, unguessable path component of the public link
	// (/c/<slug>). Unique across all tokens.
	Slug string `gorm:"uniqueIndex;not null" json:"slug"`

	// Name is the operator's label for the token (e.g. "leaked-creds-bait").
	Name string `gorm:"not null" json:"name"`

	// TargetURL is the harmless page the visitor is redirected to after the
	// hit is recorded. Must be an absolute http(s) URL.
	TargetURL string `gorm:"not null" json:"target_url"`

	// BaseURL optionally overrides the public base used to build this token's
	// link (e.g. "https://promo-event.io"). Lets the operator front the link
	// with a disposable domain so the real ForensicHub host is never exposed.
	// When empty, the global CANARY_BASE_URL config (or the request host) is used.
	BaseURL string `json:"base_url,omitempty"`

	Description string `gorm:"type:text" json:"description,omitempty"`

	// CaseID optionally ties this canary to an investigation Case, so captured
	// visits can be pivoted/attributed within that case.
	CaseID *uuid.UUID `gorm:"type:uuid;index" json:"case_id,omitempty"`

	// Active gates the link: when false the public endpoint returns 404 and
	// records nothing.
	Active bool `gorm:"not null;default:true" json:"active"`

	// CollectDetails, when true, serves a brief JS interstitial page that gathers
	// rich client data (screen, timezone, hardware, GPU, battery, network…)
	// before redirecting. When false the link is a fully transparent 302 that
	// only captures HTTP headers + IP GeoIP (the default — stealthier, opt-in
	// collection). No `default` tag: GORM omits zero-value bools that carry a
	// default, which would silently flip false→true; we always write the value.
	CollectDetails bool `gorm:"not null" json:"collect_details"`

	// RequestLocation, when true (requires CollectDetails), additionally asks the
	// browser for precise GPS coordinates via the Geolocation API — the visitor
	// sees a permission prompt and must consent.
	RequestLocation bool `gorm:"not null" json:"request_location"`

	// HitCount / LastHitAt are denormalised counters kept in sync on each hit
	// so the list view avoids a COUNT per row.
	HitCount  int        `gorm:"not null;default:0" json:"hit_count"`
	LastHitAt *time.Time `                          json:"last_hit_at,omitempty"`

	CreatedBy uuid.UUID `gorm:"type:uuid" json:"created_by"`
	CreatedAt time.Time `                 json:"created_at"`
	UpdatedAt time.Time `                 json:"updated_at"`
}

// CanaryHit is a single recorded visit to a CanaryToken's link.
type CanaryHit struct {
	ID      uuid.UUID `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	TokenID uuid.UUID `gorm:"type:uuid;not null;index"                       json:"token_id"`

	// Network identity of the visitor.
	IP           string `gorm:"index" json:"ip"`                       // best-effort real client IP
	ForwardedFor string `              json:"forwarded_for,omitempty"` // raw X-Forwarded-For chain
	RemoteAddr   string `              json:"remote_addr,omitempty"`   // direct TCP peer (proxy/CDN edge)

	// Client software, parsed best-effort from the User-Agent.
	UserAgent      string `gorm:"type:text" json:"user_agent,omitempty"`
	Browser        string `                 json:"browser,omitempty"`
	OS             string `                 json:"os,omitempty"`
	DeviceType     string `                 json:"device_type,omitempty"` // desktop | mobile | tablet | bot
	Referer        string `gorm:"type:text" json:"referer,omitempty"`
	AcceptLanguage string `                 json:"accept_language,omitempty"`

	// IsBot flags automated visits — link-preview unfurlers (Slack/Telegram/
	// Facebook/Discord/Twitter…) and security scanners (Safe Links, Proofpoint…)
	// that fetch the URL without a human clicking. Lets the UI separate real
	// human clicks from noise so the hit count stays meaningful.
	IsBot bool `gorm:"index" json:"is_bot"`

	// GeoIP enrichment (APPROXIMATE — IP-based, city level), filled asynchronously
	// after the hit via ip-api.com. Skipped for private / loopback addresses.
	Country     string  `json:"country,omitempty"`
	CountryCode string  `json:"country_code,omitempty"`
	City        string  `json:"city,omitempty"`
	ISP         string  `json:"isp,omitempty"`
	ASN         string  `json:"asn,omitempty"`
	Lat         float64 `json:"lat,omitempty"`
	Lon         float64 `json:"lon,omitempty"`

	// Network-type flags from ip-api — surface VPN/proxy, datacenter/hosting and
	// mobile-carrier connections so an analyst can judge how trustworthy the IP is.
	Mobile  bool `json:"mobile,omitempty"`
	Proxy   bool `json:"proxy,omitempty"`
	Hosting bool `json:"hosting,omitempty"`

	// ---- Rich client-side data (interstitial JS; only when CollectDetails) ----

	// Precise GPS location from the browser Geolocation API (visitor-consented).
	// Pointers so "not collected" is distinct from a real 0 value.
	GeoLat      *float64 `json:"geo_lat,omitempty"`
	GeoLon      *float64 `json:"geo_lon,omitempty"`
	GeoAccuracy *float64 `json:"geo_accuracy,omitempty"` // radius in metres
	GeoError    string   `json:"geo_error,omitempty"`    // e.g. "User denied Geolocation"

	// Device / browser fingerprint.
	Timezone     string   `json:"timezone,omitempty"`      // IANA tz (Intl)
	Languages    string   `json:"languages,omitempty"`     // navigator.languages, comma-joined
	ScreenW      int      `json:"screen_w,omitempty"`
	ScreenH      int      `json:"screen_h,omitempty"`
	ViewportW    int      `json:"viewport_w,omitempty"`
	ViewportH    int      `json:"viewport_h,omitempty"`
	PixelRatio   float64  `json:"pixel_ratio,omitempty"`
	Platform     string   `json:"platform,omitempty"`
	CPUCores     int      `json:"cpu_cores,omitempty"`     // navigator.hardwareConcurrency
	DeviceMemory float64  `json:"device_memory,omitempty"` // navigator.deviceMemory (GB)
	GPUVendor    string   `json:"gpu_vendor,omitempty"`    // WebGL UNMASKED_VENDOR
	GPURenderer  string   `json:"gpu_renderer,omitempty"`  // WebGL UNMASKED_RENDERER
	ConnType     string   `json:"conn_type,omitempty"`     // NetworkInformation effectiveType
	BatteryLevel *float64 `json:"battery_level,omitempty"` // 0..1

	// ClientData is the full raw JSON payload the interstitial reported, so any
	// field not promoted to a column above is still preserved for the UI.
	ClientData string `gorm:"type:text" json:"client_data,omitempty"`

	CreatedAt time.Time `gorm:"index" json:"created_at"`
}

File: frontend/src/api/canary.ts
import apiClient from './client'

// A canary token: an administrator-created tracking link. When its public link
// (/c/<slug>) is opened, the server records the visitor and 302-redirects to a
// benign target chosen by the operator.
export interface CanaryToken {
  id: string
  slug: string
  name: string
  target_url: string
  base_url?: string
  description?: string
  active: boolean
  collect_details: boolean
  request_location: boolean
  case_id?: string | null
  hit_count: number
  unique_visitors: number
  last_hit_at?: string | null
  created_by: string
  created_at: string
  updated_at: string
  link: string // fully-qualified public URL, built by the server
}

// A single recorded visit to a canary link.
export interface CanaryHit {
  id: string
  token_id: string
  ip: string
  forwarded_for?: string
  remote_addr?: string
  user_agent?: string
  browser?: string
  os?: string
  device_type?: string
  is_bot?: boolean
  referer?: string
  accept_language?: string
  country?: string
  country_code?: string
  city?: string
  isp?: string
  asn?: string
  lat?: number
  lon?: number
  mobile?: boolean
  proxy?: boolean
  hosting?: boolean
  // Precise GPS (browser Geolocation API, with consent)
  geo_lat?: number
  geo_lon?: number
  geo_accuracy?: number
  geo_error?: string
  // Device / browser fingerprint (interstitial JS)
  timezone?: string
  languages?: string
  screen_w?: number
  screen_h?: number
  viewport_w?: number
  viewport_h?: number
  pixel_ratio?: number
  platform?: string
  cpu_cores?: number
  device_memory?: number
  gpu_vendor?: string
  gpu_renderer?: string
  conn_type?: string
  battery_level?: number
  client_data?: string // full raw JSON dump
  created_at: string
}

export interface CreateCanaryToken {
  name: string
  target_url: string
  description?: string
  slug?: string // optional custom slug; random if omitted
  base_url?: string // optional link domain; falls back to server default
  case_id?: string // optional Case to attach to
  collect_details?: boolean // serve JS interstitial to gather rich client data
  request_location?: boolean // also prompt for precise GPS
}

export interface BulkCreateCanaryTokens {
  targets: string[] // one redirect URL per link
  name_prefix?: string // optional; default label = target host
  base_url?: string
  description?: string
  collect_details?: boolean
  request_location?: boolean
}

export interface BulkCreateResult {
  created: CanaryToken[]
  errors: { target: string; error: string }[]
}

export interface UpdateCanaryToken {
  name?: string
  target_url?: string
  description?: string
  active?: boolean
  base_url?: string
  case_id?: string | null // "" / null detaches; a uuid attaches
  collect_details?: boolean
  request_location?: boolean
}

export const canaryApi = {
  async list(): Promise<CanaryToken[]> {
    const { data } = await apiClient.get('/canary/tokens')
    return data.data
  },

  async create(input: CreateCanaryToken): Promise<CanaryToken> {
    const { data } = await apiClient.post('/canary/tokens', input)
    return data.data
  },

  async createBulk(input: BulkCreateCanaryTokens): Promise<BulkCreateResult> {
    const { data } = await apiClient.post('/canary/tokens/bulk', input)
    return data.data
  },

  async update(id: string, input: UpdateCanaryToken): Promise<CanaryToken> {
    const { data } = await apiClient.patch(`/canary/tokens/${id}`, input)
    return data.data
  },

  async remove(id: string): Promise<void> {
    await apiClient.delete(`/canary/tokens/${id}`)
  },

  async hits(id: string): Promise<CanaryHit[]> {
    const { data } = await apiClient.get(`/canary/tokens/${id}/hits`)
    return data.data
  },

  async clearHits(id: string): Promise<void> {
    await apiClient.delete(`/canary/tokens/${id}/hits`)
  },

  // Pivot a captured visitor IP into an OSINT scan; returns the new scan id.
  async scanHit(tokenId: string, hitId: string): Promise<string> {
    const { data } = await apiClient.post(`/canary/tokens/${tokenId}/hits/${hitId}/scan`)
    return data.data.scan_id
  },
}
